---
title: "CLL_scRNASeq_alt_filter"
author: "ASG"
date: "7/20/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## control and treatment cells alternative filtering

Based on the QC metrics of the data (see CLL_scRNASeq_seurat.Rmd, assessing QC metrics section), seems obvious that the two conditions have distinct cell populations based on cell count, transcripts per cell, genes per cell, genes versus UMI counts and complexity. All these metrics are distinct for the two conditions and filtering out cells as proposed in that analysis will filter out close to half the population of cells in the control condition.

In this analysis, we will using different criteria for filtering

## loading needed packages
```{r include=FALSE}
library(Seurat)
library(tidyverse)
library(Matrix)
library(patchwork)
library(RCurl)
library(cowplot)
```


## reading in data
The data provided is in the "data" folder in this project. Consists of 2 sub-folders representing the two conditions. 

```{r}
# load the cll_control data
cll_cntrl_data <- Read10X("./data/control/con1-filtered_feature_bc_matrix/")
# initializing the Seurat object with raw control data
cll_control <- CreateSeuratObject(counts = cll_cntrl_data, project = "cll_cntrl", min.cells = 3, min.features = 200)
cll_control
```

```{r}
# load the cll_treatment data
cll_treat1_data <- Read10X("./data/treat1/treat1-filtered_feature_bc_matrix/")
# initializing the Seurat object with raw treatment data (non-normalized)
cll_treatment <- CreateSeuratObject(counts = cll_treat1_data, project = "cll_treatment", min.cells = 3, min.features = 200)
cll_treatment
```
### adding gene annotations
This is required for later annotating identified markers
We downloaded the annotations file from the HBCtraining link
```{r}
annotations <- read.csv("./data/annotation.csv")
```

## Merging both seurat objects, QC visualization and filtering
```{r}
# merging both seurat objects
merged_seurat <- merge(x = cll_control,
                       y = cll_treatment,
                       add.cell.id = c("cntrl", "treat"))
```

## Some data munging
```{r}
# explore merged data
View(merged_seurat@meta.data)
```

there are 3 columns in the current object, we add two additional columns : 
number of genes detected per UMI: this metric gives us an idea of the complexity of our dataset (more genes detected per UMI, more complex our data)
mitochondrial ratio: this metric will give us a percentage of cell reads originating from the mitochondrial genes
```{r}
# add number of genes per UMI
merged_seurat$log10GenesPerUMI <- log10(merged_seurat$nFeature_RNA) /
  log10(merged_seurat$nCount_RNA)
```

```{r}
# add the ratio of mitochondrial genes present
merged_seurat$mitoRatio <- PercentageFeatureSet(object = merged_seurat, pattern = "^MT-")
merged_seurat$mitoRatio <- merged_seurat$mitoRatio / 100
```

### additional metadata columns for convenience
```{r}
# extracting metadata as a dataframe
metadata <- merged_seurat@meta.data
```

adding a new column for cell identifiers

This information is currently located in the row names of our metadata dataframe. We will keep the rownames as is and duplicate it into a new column called cells.
```{r}
# add cell IDs to metadata
metadata$cells <- rownames(metadata)
```

Each cell ID has a cntrl_ or treat_ prefix as we had specified when we merged the Seurat objects. We can use this prefix to create a new column indicating which condition each cell is classified under. We will call this column sample:
```{r}
# create a column called sample
metadata$sample <- NA
metadata$sample[which(str_detect(metadata$cells, "^cntrl_"))] <- "ctrl"
metadata$sample[which(str_detect(metadata$cells, "^treat_"))] <- "treat"
```

rename some of the existing columns in metadata to be more intuitive:
```{r}
# Rename columns
metadata <- metadata %>% 
  rename(seq_folder = orig.ident,
         nUMI = nCount_RNA,
         nGene = nFeature_RNA)
```

### saving new metadata into the merged seurat object
Before we assess our metrics we are going to save all of the work we have done thus far back into our Seurat object. We simply assign the dataframe into the meta.data slot:
```{r}
# add metadata back to the merged seurat object
merged_seurat@meta.data <- metadata

# create a .RData object to load back at any time
# save(merged_seurat, file = "./data/cll_merged_filtered_seurat.RData") # commented out since already saved on macbookair
```


## Cell filtering
Based on above analysis and results from "CLL_separate_analysis_ctrl_treat.Rmd", we will use foll; thresholds:
500> nUMI < 32000, # see reasoning in "CLL_separate_analysis_ctrl_treat.Rmd"
nGene > 300,
mitoRatio < 0.2, # see reasoning in "CLL_separate_analysis_ctrl_treat.Rmd"
log10GeneperUMI > 0.8

```{r}
# filtering out low quality cells based on selected thresholds
filtered_seurat <- subset(x = merged_seurat,
                          subset = (nUMI > 500 &
                                     nUMI < 32000) & 
                            (nGene >= 300) & 
                            (log10GenesPerUMI > 0.8) &
                            (mitoRatio < 0.2))
```

## Gene-level filtering

Within our data we will have many genes with zero counts. These genes can dramatically reduce the average expression for a cell and so we will remove them from our data. We will start by identifying which genes have a zero count in each cell:
```{r}
# Extract counts
counts <- GetAssayData(object = filtered_seurat, slot = "counts")

# Output a logical matrix specifying for each gene on whether or not there are more than zero counts per cell
nonzero <- counts > 0
```

Now, we will perform some filtering by prevalence. If a gene is only expressed in a handful of cells, it is not particularly meaningful as it still brings down the averages for all other cells it is not expressed in. For our data we choose to keep only genes which are expressed in 10 or more cells. By using this filter, genes which have zero counts in all cells will effectively be removed.

```{r}
# Sums all TRUE values and returns TRUE if more than 10 TRUE values per gene
keep_genes <- Matrix::rowSums(nonzero) >= 10

# Only keeping those genes expressed in more than 10 cells
filtered_counts <- counts[keep_genes, ]
```

Finally, take those filtered counts and create a new Seurat object for downstream analysis.

```{r}
# Reassign to filtered Seurat object
filtered_seurat <- CreateSeuratObject(filtered_counts, meta.data = filtered_seurat@meta.data)
```


## Reassessing QC metrics
```{r}
# extracting metadata as a dataframe from the filtered_seurat object
filtered_metadata <- filtered_seurat@meta.data
```


```{r}
# Visualize the number of cells in each sample
  filtered_metadata %>% 
  ggplot() +
  geom_bar(aes(sample, fill = sample)) +
  theme_light() +
  theme(plot.title = element_text(hjust = 0.5, face = "bold")) +
  ggtitle("NCells_filtered")

```
After filtering, the number of ctrl and treated cells is
```{r}
nrow(filtered_metadata %>% filter(sample == "ctrl"))
nrow(filtered_metadata %>% filter(sample == "treat"))
```

```{r}
# visualize the number of transcripts (UMIs) per cell
filtered_metadata %>% 
  ggplot(aes(x = nUMI, color = sample, fill = sample)) +
  geom_density(alpha = 2/10) +
  theme_light() + 
  scale_x_log10() + 
  ylab("cell density") + 
  geom_vline(xintercept = 500) +
  theme(plot.title = element_text(hjust = 0.5, face = "bold")) + 
  ggtitle("transcripts per cell after filtering")
```
```{r}
# visualize number of genes (nGene) per cell with a histogram
filtered_metadata %>% 
  ggplot(aes(nGene, color = sample, fill = sample)) +
  geom_density(alpha = 2/10) + 
  theme_light() + 
  scale_x_log10() + 
  geom_vline(xintercept = 300) + 
  theme(plot.title = element_text(hjust = 0.5, face = "bold")) + 
  ggtitle("genes per cell after filtering")
```
```{r}
# visualize number of genes as a boxplot
filtered_metadata %>% 
  ggplot(aes(x = sample, y = log10(nGene), fill = sample)) + 
  geom_boxplot(alpha = 5/10) + 
  theme_light() + 
  theme(plot.title = element_text(hjust = 0.5, face = "bold")) + 
  ggtitle("genes per cell after filtering")
```
```{r}
# scatter plot of UMIs versus genes detected per cell colored by ratio of mitochondrial genes
filtered_metadata %>% 
  ggplot(aes(x = nUMI, y = nGene, color = mitoRatio)) +
  geom_jitter() +
  scale_color_gradient(low = "gray90", high = "black") +
  stat_smooth(method = lm) +
  scale_x_log10() +
  scale_y_log10() +
  theme_light() + 
  facet_wrap(~sample)
```

```{r}
# visualizing mitochondrial genes per cell 
filtered_metadata %>% 
  ggplot(aes(mitoRatio, color = sample, fill = sample)) + 
  geom_density(alpha = 2/10) + 
  theme_light() + 
  scale_x_log10() + 
  geom_vline(xintercept = 0.2)
  theme(plot.title = element_text(hjust = 0.5, face = "bold")) + 
  ggtitle("mitochondrial gene count ratio per cell after filtering")
```

```{r}
# visualizing genes per UMI to assess complexity
filtered_metadata %>% 
  ggplot(aes(log10GenesPerUMI, color = sample, fill = sample)) + 
  geom_density(alpha = 2/10) + 
  theme_light()+
  scale_x_log10() + 
  geom_vline(xintercept = 0.8)
  theme(plot.title = element_text(hjust = 0.5, face = "bold")) + 
  ggtitle("genes per UMI after filtering")
```

## Assigning cell cycle phases
#### Assigning cell cycle phases and scores
To assign each cell a score based on its expression of G2/M and S phase markers, we can use the Seurat function CellCycleScoring(). This function calculates cell cycle phase scores based on canonical markers that required as input.

```{r}
# load cell cycle data; the RData file was downloaded from the seurat github site; the HBC tutorial github data folder does not have any .rda files!
load("./data/cc.genes.updated.2019.rda")

# scoring cells for cell cycle
filtered_phased_seurat <- CellCycleScoring(filtered_seurat,
                                 s.features = cc.genes.updated.2019$s.genes,
                                 g2m.features = cc.genes.updated.2019$g2m.genes)
```

```{r}
# view cell cycle scores and phases assigned to cells
glimpse(filtered_phased_seurat@meta.data)
```


## Using sctransform

### From the vignette on sctransform from Hafemeister & Satija
Using sctransform in Seurat
Christoph Hafemeister & Rahul Satija

Compiled: 2021-06-14

Source: vignettes/sctransform_vignette.Rmd
Biological heterogeneity in single-cell RNA-seq data is often confounded by technical factors including sequencing depth. The number of molecules detected in each cell can vary significantly between cells, even within the same celltype. Interpretation of scRNA-seq data requires effective pre-processing and normalization to remove this technical variability. In Hafemeister and Satija, 2019 we introduce a modeling framework for the normalization and variance stabilization of molecular count data from scRNA-seq experiment. This procedure omits the need for heuristic steps including pseudocount addition or log-transformation and improves common downstream analytical tasks such as variable gene selection, dimensional reduction, and differential expression.

In this vignette, we demonstrate how using sctransform based normalization enables recovering sharper biological distinction compared to log-normalization.

library(Seurat)
library(ggplot2)
library(sctransform)



### Additional note regarding omitting NormalizeData, ScaleData and FindVariableFeatures functions
Apply sctransform normalization

Note that this single command replaces NormalizeData(), ScaleData(), and FindVariableFeatures().
Transformed data will be available in the SCT assay, which is set as the default after running sctransform
During normalization, we can also remove confounding sources of variation, for example, mitochondrial mapping percentage

### applying sctransform directly to filtered_phased_seurat object

We will apply sctransform directly to the filtered_phased_seurat object without normalizing etc.

During first attempt to assign identities to clusters, we noticed some clusters are only distinguished by ribosomal protein and cell cycle markers. Hence we need to regress out cell cycle effects on the data. We start at assigning cell cycle phases to cells after the QC metrics section and then regressing out the cell cycle during the sctransform procedure. 

We will then redo the downstream analysis after the sctransform procedure.

We will also regress out the mitoRatio factor as done in the "redo" sections of the earlier analysis. This is because, we re-did the entire analysis without regressing out the mitochondrial contribution and found than one cluster amongst 14 (cluster 4 at resolution of 0.8) had only mitochondrial genes in the top 10.

```{r}
# split seurat object by condition for sctransform
filtered_phased_seurat_noMito <- SplitObject(filtered_phased_seurat,split.by = "sample")
filtered_phased_seurat_noMito <- filtered_phased_seurat_noMito[c("ctrl", "treat")]
```

SCTransform can generate large objects in terms of memory, hence adjust the limit for allowable object sizes within R 
```{r}
options(future.globals.maxSize = 4000 * 1024^2)
```

Iterating sctransform over each object in split_filtered_seurat
```{r}
 for (i in seq_along(filtered_phased_seurat_noMito)){
  filtered_phased_seurat_noMito[[i]] = SCTransform(filtered_phased_seurat_noMito[[i]], vars.to.regress = c("mitoRatio", "S.Score", "G2M.Score")) # regressing out mitochondrial gene ratio and cell cycle scores as factors
}
```

checking assays stored in objects
```{r}
filtered_phased_seurat_noMito$ctrl@assays
```
```{r}
filtered_phased_seurat_noMito$treat@assays
```


## Integration

goal is to integrate across conditions(control and treatment) by aligning similar cell types

### Specifying 3000 most variable genes to be used for integration as identified by sctransform instead of 2000 
```{r}
# select most variable features to be used for integration
integ_features <- SelectIntegrationFeatures(object.list = filtered_phased_seurat_noMito,
                                            nfeatures = 3000)
```

### prepare SCTransform object for integration
```{r}
# prepare the SCT list object for integration
filtered_phased_seurat_noMito <- PrepSCTIntegration(object.list = filtered_phased_seurat_noMito,
                                   anchor.features = integ_features)
```

### perform CCA, find best buddies / anchors, filter incorrect anchors
```{r}
# find best buddies / anchors
integ_anchors <- FindIntegrationAnchors(object.list = filtered_phased_seurat_noMito,
                                        normalization.method = "SCT",
                                        anchor.features = integ_features)
```
### Integrate across conditions
```{r}
# integrate across conditions
integrated_filtered_phased_noMito_seurat <- IntegrateData(anchorset = integ_anchors,
                                   normalization.method = "SCT")
```
### Visualizing the integrated data with PCA
```{r}
# PCA
integrated_filtered_phased_noMito_seurat <- RunPCA(object = integrated_filtered_phased_noMito_seurat)

```
```{r}
# plot PCA
PCAPlot(object = integrated_filtered_phased_noMito_seurat,
        split.by = "sample")
```

### Visualizing with UMAP
```{r}
# run UMAP
integrated_filtered_phased_noMito_seurat <- RunUMAP(integrated_filtered_phased_noMito_seurat,
                             dims = 1:40,
                             reduction = "pca")
```

```{r}
# plot UMAP
DimPlot(integrated_filtered_phased_noMito_seurat)

```
```{r}
#visualizing side by side with UMAP
DimPlot(integrated_filtered_phased_noMito_seurat,
        split.by = "sample")
```
### Saving the integrated seurat object in results
```{r}
# save integrated seurat object
# saveRDS(integrated_filtered_phased_noMito_seurat, "./results/integrated_filtered_phased_noMito_seurat.rds") # saved in macbookair
```

## Clustering
### Determining number of PCs to use
```{r}
# elbow / scree plot
ElbowPlot(integrated_filtered_phased_noMito_seurat,
          ndims = 50)
```
We will use 50 PCs to determine clusters

### k-nearest neighbors, clusters at various resolutions
```{r}
# K-nearest neighbors graph
integrated_filtered_phased_noMito_seurat <- FindNeighbors(object = integrated_filtered_phased_noMito_seurat,
                                   dims = 1:50)

#determine clusters at various resolutions
integrated_filtered_phased_noMito_seurat <- FindClusters(object = integrated_filtered_phased_noMito_seurat,
                                  resolution = c(0.4, 0.5,0.6, 0.8, 1.0, 1.4))
```

### looking at various resolutions
```{r}
integrated_filtered_phased_noMito_seurat@meta.data %>% 
  View()
```

Choosing resolution of 1.4
```{r}
# assigning identity of clusters
Idents(object = integrated_filtered_phased_noMito_seurat) <- "integrated_snn_res.1.4"
```

We have calculated UMAP earlier, hence we only visualize the clusters
```{r}
# Plot the UMAP
DimPlot(object = integrated_filtered_phased_noMito_seurat,
        reduction = "umap",
        label = TRUE,
        label.size = 6)
```
plotting a split UMAP plot
```{r}
# plot UMAP by condition
DimPlot(object = integrated_filtered_phased_noMito_seurat,
        split.by = "sample",
        reduction = "umap",
        label = TRUE,
        label.size = 5)
```
Choosing resolution of 1.0
```{r}
# assigning identity of clusters
Idents(object = integrated_filtered_phased_noMito_seurat) <- "integrated_snn_res.1.0"
```

We have calculated UMAP earlier, hence we only visualize the clusters
```{r}
# Plot the UMAP
DimPlot(object = integrated_filtered_phased_noMito_seurat,
        reduction = "umap",
        label = TRUE,
        label.size = 6)
```

plotting a split UMAP plot
```{r}
# plot UMAP by condition
DimPlot(object = integrated_filtered_phased_noMito_seurat,
        split.by = "sample",
        reduction = "umap",
        label = TRUE,
        label.size = 5)
```
Choosing resolution of 0.8
```{r}
# assigning identity of clusters
Idents(object = integrated_filtered_phased_noMito_seurat) <- "integrated_snn_res.0.8"
```

We have calculated UMAP earlier, hence we only visualize the clusters
```{r}
# Plot the UMAP
DimPlot(object = integrated_filtered_phased_noMito_seurat,
        reduction = "umap",
        label = TRUE,
        label.size = 6)
```

plotting a split UMAP plot
```{r}
# plot UMAP by condition
DimPlot(object = integrated_filtered_phased_noMito_seurat,
        split.by = "sample",
        reduction = "umap",
        label = TRUE,
        label.size = 6)
```

Choosing resolution of 0.6
```{r}
# assigning identity of clusters
Idents(object = integrated_filtered_phased_noMito_seurat) <- "integrated_snn_res.0.6"
```

We have calculated UMAP earlier, hence we only visualize the clusters
```{r}
# Plot the UMAP
DimPlot(object = integrated_filtered_phased_noMito_seurat,
        reduction = "umap",
        label = TRUE,
        label.size = 6)
```
plotting a split UMAP plot
```{r}
# plot UMAP by condition
DimPlot(object = integrated_filtered_phased_noMito_seurat,
        split.by = "sample",
        reduction = "umap",
        label = TRUE,
        label.size = 6)
```

Choosing resolution of 0.5
```{r}
# assigning identity of clusters
Idents(object = integrated_filtered_phased_noMito_seurat) <- "integrated_snn_res.0.5"
```

We have calculated UMAP earlier, hence we only visualize the clusters
```{r}
# Plot the UMAP
DimPlot(object = integrated_filtered_phased_noMito_seurat,
        reduction = "umap",
        label = TRUE,
        label.size = 6)
```
plotting a split UMAP plot
```{r}
# plot UMAP by condition
DimPlot(object = integrated_filtered_phased_noMito_seurat,
        split.by = "sample",
        reduction = "umap",
        label = TRUE,
        label.size = 6)
```
Choosing resolution of 0.4
```{r}
# assigning identity of clusters
Idents(object = integrated_filtered_phased_noMito_seurat) <- "integrated_snn_res.0.4"
```

We have calculated UMAP earlier, hence we only visualize the clusters
```{r}
# Plot the UMAP
DimPlot(object = integrated_filtered_phased_noMito_seurat,
        reduction = "umap",
        label = TRUE,
        label.size = 6)
```
plotting a split UMAP plot
```{r}
# plot UMAP by condition
DimPlot(object = integrated_filtered_phased_noMito_seurat,
        split.by = "sample",
        reduction = "umap",
        label = TRUE,
        label.size = 6)
```
Settling on 11 clusters with resolution of 0.5.

```{r}
# assigning identity of clusters
Idents(object = integrated_filtered_phased_noMito_seurat) <- "integrated_snn_res.0.5"
```

plotting a split UMAP plot
```{r}
# plot UMAP by condition
DimPlot(object = integrated_filtered_phased_noMito_seurat,
        split.by = "sample",
        reduction = "umap",
        label = TRUE,
        label.size = 6)
```

## QC for Clustering
### Distribution of cells per cluster in each condition
```{r}
# extracting identity and sample information to determine cells per cluster in each condition
n_cells <- FetchData(integrated_filtered_phased_noMito_seurat,
                     vars = c("ident", "orig.ident")) %>% 
  count(ident, orig.ident) %>% 
  spread(ident, n)

n_cells
```

```{r}
# visualizing by count and condition
DimPlot(object = integrated_filtered_phased_noMito_seurat,
        split.by = "sample",
        reduction = "umap",
        label = TRUE) +
  NoLegend()
```
### Segregation of clusters by other uninteresting sources of variation
Exploring additional metrics, such as the number of UMIs and genes per cell, S-phase and G2M-phase markers, and mitochondrial gene expression by UMAP. Looking at the individual S and G2M scores can give us additional information to checking the phase as we did previously.
```{r}
# Plot metrics present in seurat_integrated@meta.data
metrics <- c("nUMI", "nGene","S.Score", "G2M.Score", "mitoRatio")

FeaturePlot(integrated_filtered_phased_noMito_seurat,
            reduction = "umap",
            features = metrics,
            pt.size = 0.4,
            order = TRUE,
            min.cutoff = "q10",
            label = TRUE)
```
Some indication that cluster 3 may be clustering on basis of nUMIs and nGenes. Will need more looking into?


### Exploring PCs that drive different clusters
The hope is that defined PCs separate the cell types well. To do this, we extract UMAP coordinate information for the cells along with the corresponding scores for each of the first 16 PCs to view by UMAP
```{r}
# defining information of interest in the seurat object
columns <- c(paste0("PC_", 1:16),
             "ident",
             "UMAP_1", "UMAP_2")
# extracting this information from the seurat object
pc_data <- FetchData(integrated_filtered_phased_noMito_seurat, 
                     vars = columns)
```

In the UMAP plots below, cells are colored by their score for each of first 16 PCs
```{r}
# adding cluster label to center of cluster on UMAP
umap_label <- FetchData(integrated_filtered_phased_noMito_seurat,
                        vars = c("ident", "UMAP_1", "UMAP_2")) %>% 
  group_by(ident) %>% 
  summarize(x = mean(UMAP_1), y = mean(UMAP_2))

# plotting a UMAP plot of each of first 16 PCs
map(paste0("PC_", 1:16), function(pc){
  ggplot(pc_data,
         aes(UMAP_1, UMAP_2)) +
    geom_point(aes_string(color = pc),
               alpha = 0.7) + 
    scale_color_gradient(guide = "none",
                         low = "grey90",
                         high = "blue") + 
    geom_text(data = umap_label,
              aes(label = ident, x, y)) + 
    ggtitle(pc)
}) %>% 
  plot_grid(plotlist = .) 
```

Examine PCA results
```{r}
print(integrated_filtered_phased_noMito_seurat[["pca"]], dims = 1:5, nfeatures = 5)
```

## Visualizing known cell-type markers

```{r}
# select the RNA counts slot to be default assay
DefaultAssay(integrated_filtered_phased_noMito_seurat) <- "RNA"

# Normalize RNA data for visualization
integrated_filtered_phased_noMito_seurat <- NormalizeData(integrated_filtered_phased_noMito_seurat, verbose = FALSE)
```

### Plotting known immune cell markers
```{r}
# DotPlot using most variable features and other known markers
DotPlot(integrated_filtered_phased_noMito_seurat,features = c("CD79A","MS4A1", "CD19","CD74","CD44", "ID3", "ACTB","CCR7","MIR155HG","MIR210HG","FOSB","IFIT3", "SNHG25"))
```
Major T cell (CD3D, CD4, CD8A and IL7R) markers were not found in the data
## Identity of clusters
Based on several gene transcripts expressed, as seen above, all clusters are likely B-lymphocytes or pre-B lymphocytes. Further classification of clusters requires specialized knowledge. 
Based on low expression of most markers plotted above cluster 6 may show biological significance.

## Finding markers differentially expressed between B cells and other clusters

### DE between cluster6 (ribosomal proteins) and 0,1,2,3
```{r}
# DE between 0,1,2, 3 and cluster 6
cluster6cells <- FindMarkers(integrated_filtered_phased_noMito_seurat,
                             ident.1 = 6,
                             ident.2 = c(0,1,2,3))
# add gene symbols to table
cluster6cells <- cluster6cells %>%
  rownames_to_column(var = "gene") %>%
  left_join(y = unique(annotations[, c("gene_name", "description")]),
             by = c("gene" = "gene_name"))
# reorder columns
cluster6cells <- cluster6cells[ ,c(1, 3:5, 2, 6:7)]

# sort by padj value
cluster6cells <-  cluster6cells %>% 
  arrange(p_val_adj)

View(cluster6cells)
```

Of the top 50 genes differentially expressed in cluster 6 all except MALAT1 (#10), RACK1 (#41) and HLA-DRA (#47) are ribosomal proteins indicating cluster 6 cells differ only by metabolic status.



### DE between cluster10 (ribosomal proteins) and 0,1,2
```{r}
# DE between 0,1,2 and cluster 10
cluster10cells <- FindMarkers(integrated_filtered_phased_noMito_seurat,
                             ident.1 = 10,
                             ident.2 = c(0,1,2,3))
# add gene symbols to table
cluster10cells <- cluster10cells %>%
  rownames_to_column(var = "gene") %>%
  left_join(y = unique(annotations[, c("gene_name", "description")]),
             by = c("gene" = "gene_name"))
# reorder columns
cluster10cells <- cluster10cells[ ,c(1, 3:5, 2, 6:7)]

# sort by padj value
cluster10cells <-  cluster10cells %>% 
  arrange(p_val_adj)

View(cluster10cells)
```
Cluster 10 has mostly ribosomal proteins showing fold changes, with # 1 gene being SNHG25 which is of uncertain significance.


## Cluster identities
```{r}
# assigning cluster identities
integrated_ID_seurat <- integrated_filtered_phased_noMito_seurat
integrated_ID_seurat <- RenameIdents(object = integrated_ID_seurat, 
                               "0" = "Bcell_CD19+_0",
                               "1" = "Bcell_CD19+_1",
                               "2" = "Bcell_CD19+_2",
                               "3" = "Bcell_CD19+_3",
                               "4" = "Bcell_CD19+_MIR155HG",
                               "5" = "Bcell_CD19+_MIR210HG",
                               "6" = "CD19_low_CCR7_high",
                               "7" = "Bcell_CD19+_7",
                               "8" = "CD44_high_FOSB",
                               "9" = "multi_IF_induced",
                               "10" = "SNHG25+"
                               )

# Plot the UMAP
DimPlot(object = integrated_ID_seurat, 
        reduction = "umap", 
        label = TRUE)
```
## Save final R object
```{r}
# Save final R object
write_rds(integrated_ID_seurat,
          path = "results/seurat_labelled_final_output.rds")

# Create and save a text file with sessionInfo
sink("sessionInfo_scrnaseq_07272021.txt")
sessionInfo()
sink()
```

## Get all markers
```{r}
# ensure default assay is "RNA" for detecting markers
DefaultAssay(integrated_filtered_phased_noMito_seurat) <- "RNA"
```

```{r}
# get_all_markers function
all_markers <- FindAllMarkers(object = integrated_filtered_phased_noMito_seurat)

all_markers <-  all_markers %>% 
  left_join(y = unique(annotations[ , c("gene_name", "description")]),
              by = c("gene" = "gene_name"))
```

### Evaluating the marker genes to identify clusters

```{r}
# extract top 10 markers per cluster
top10 <- all_markers %>% 
  group_by(cluster) %>% 
  top_n(n = 10,
        wt = avg_log2FC)

View(top10)
```

## Without Integration per sctransform pipeline from Satijalab
As laid out here
https://satijalab.org/seurat/articles/sctransform_vignette.html

### Splitting the merged conditions into 2 separate seurat objects
```{r}
filtered_ctrl_seurat <- split_filtered_seurat_noMito$ctrl
filtered_treat_seurat <- split_filtered_seurat_noMito$treat
```

### run PCA and UMAP on ctrl
```{r}
ctrl <- RunPCA(filtered_ctrl_seurat,
               verbose = FALSE)
ctrl <- RunUMAP(ctrl,
                dims = 1:40,
                verbose = FALSE)
ctrl <- FindNeighbors(ctrl,
                      dims = 1:40,
                      verbose = FALSE)
ctrl <- FindClusters(ctrl,
                     verbose = FALSE)
DimPlot(ctrl, label = TRUE) + NoLegend()
```
### Violin plots of top 10 variable features in ctrl noted in sctransform output
```{r}
VlnPlot(ctrl,
        features = c("IGLC1", "HSPA5", "NMB", "MT-CO1", "ID3", "IL4I1", "IGFBP7", "IFIT3", "SNHG25","MT-CO3"), pt.size = 0.2, ncol = 4)
```


### Run PCA and UMAP on treat
```{r}
treat <- RunPCA(filtered_treat_seurat,
               verbose = FALSE)
treat <- RunUMAP(treat,
                dims = 1:40,
                verbose = FALSE)
treat <- FindNeighbors(treat,
                      dims = 1:40,
                      verbose = FALSE)
treat <- FindClusters(treat,
                     verbose = FALSE)
DimPlot(treat, label = TRUE) + NoLegend()

```

Violin plots of top 10 variable features in treat noted in sctransform output
```{r}
VlnPlot(treat,
        features = c("STMN1", "UBE2C", "LGALS1", "ACTB", "MKI67", "GZMB", "TYMS", "MT2A", "TOP2A",
"MT-CO1" ), pt.size = 0.2, ncol = 4)
```