---
title: "CLL_separate_analysis_30thaug2021"
author: "ASG"
date: "8/30/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Libraries
We check what packages are loaded
```{r}
# already loaded
(.packages())
```
If needed we load:

```{r}
# libraries needed
library(tidyverse)
library(skimr)
library(RCurl)
library(cowplot)
library(Matrix)
library(patchwork)
library(Seurat)
library(sctransform)
```

## Reasoning

This is an analysis of ctrl and treatment data done separately without integration. 

An imp. goal of the experiment is to successfully culture CLL cells in vitro, which has not been successfully carried out so far. 

An additional goal could be to mimic the lymph node microenvironment in which CLL cells are known to actively proliferate, where as they remain quiescent in peripheral blood. 

This analysis attempts to identify the differences in CLL cell transcriptome between the two conditions and possibly trace the trajectory of development of CLL cells.

## reading in data
The data provided is in the "data" folder in this project. Consists of 2 sub-folders representing the two conditions. 

```{r}
# load the cll_control data
cll_cntrl_data <- Read10X("./data/control/con1-filtered_feature_bc_matrix/")
# initializing the Seurat object with raw control data
cll_control <- CreateSeuratObject(counts = cll_cntrl_data, project = "cll_cntrl", min.cells = 3, min.features = 200)
cll_control
```

```{r}
# load the cll_treatment data
cll_treat1_data <- Read10X("./data/treat1/treat1-filtered_feature_bc_matrix/")
# initializing the Seurat object with raw treatment data (non-normalized)
cll_treatment <- CreateSeuratObject(counts = cll_treat1_data, project = "cll_treatment", min.cells = 3, min.features = 200)
cll_treatment
```
### adding gene annotations
This is required for later annotating identified markers
We downloaded the annotations file from the HBCtraining link
```{r}
annotations <- read.csv("./data/annotation.csv")
```


## Merging both seurat objects, QC visualization and filtering
```{r}
# merging both seurat objects
merged_seurat <- merge(x = cll_control,
                       y = cll_treatment,
                       add.cell.id = c("cntrl", "treat"))
```

## Some data munging
```{r}
# explore merged data
View(merged_seurat@meta.data)
```

there are 3 columns in the current object, we add two additional columns : 
number of genes detected per UMI: this metric gives us an idea of the complexity of our dataset (more genes detected per UMI, more complex our data)
mitochondrial ratio: this metric will give us a percentage of cell reads originating from the mitochondrial genes
```{r}
# add number of genes per UMI
merged_seurat$log10GenesPerUMI <- log10(merged_seurat$nFeature_RNA) /
  log10(merged_seurat$nCount_RNA)
```

```{r}
# add the ratio of mitochondrial genes present
merged_seurat$mitoRatio <- PercentageFeatureSet(object = merged_seurat, pattern = "^MT-")
merged_seurat$mitoRatio <- merged_seurat$mitoRatio / 100
```

### additional metadata columns for convenience
```{r}
# extracting metadata as a dataframe
metadata <- merged_seurat@meta.data
```

adding a new column for cell identifiers

This information is currently located in the row names of our metadata dataframe. We will keep the rownames as is and duplicate it into a new column called cells.
```{r}
# add cell IDs to metadata
metadata$cells <- rownames(metadata)
```

Each cell ID has a cntrl_ or treat_ prefix as we had specified when we merged the Seurat objects. We can use this prefix to create a new column indicating which condition each cell is classified under. We will call this column sample:
```{r}
# create a column called sample
metadata$sample <- NA
metadata$sample[which(str_detect(metadata$cells, "^cntrl_"))] <- "ctrl"
metadata$sample[which(str_detect(metadata$cells, "^treat_"))] <- "treat"
```

rename some of the existing columns in metadata to be more intuitive:
```{r}
# Rename columns
metadata <- metadata %>% 
  rename(seq_folder = orig.ident,
         nUMI = nCount_RNA,
         nGene = nFeature_RNA)
```

### saving new metadata into the merged seurat object
Before we assess our metrics we are going to save all of the work we have done thus far back into our Seurat object. We simply assign the dataframe into the meta.data slot:
```{r}
# add metadata back to the merged seurat object
merged_seurat@meta.data <- metadata

# create a .RData object to load back at any time
 # save(merged_seurat, file = "./data/cll_merged_seurat.RData") # commented out since already saved on macbookair
```


## Cell filtering
Based on above analysis and results from "CLL_separate_analysis_ctrl_treat.Rmd", we will use foll; thresholds:
500> nUMI < 32000, # see reasoning in "CLL_separate_analysis_ctrl_treat.Rmd"
nGene > 300,
mitoRatio < 0.2, # see reasoning in "CLL_separate_analysis_ctrl_treat.Rmd"
log10GeneperUMI > 0.8

```{r}
# filtering out low quality cells based on selected thresholds
filtered_seurat <- subset(x = merged_seurat,
                          subset = (nUMI > 500 &
                                     nUMI < 32000) & 
                            (nGene >= 300) & 
                            (log10GenesPerUMI > 0.8) &
                            (mitoRatio < 0.2))
```

## Gene-level filtering

Within our data we will have many genes with zero counts. These genes can dramatically reduce the average expression for a cell and so we will remove them from our data. We will start by identifying which genes have a zero count in each cell:
```{r}
# Extract counts
counts <- GetAssayData(object = filtered_seurat, slot = "counts")

# Output a logical matrix specifying for each gene on whether or not there are more than zero counts per cell
nonzero <- counts > 0
```

Now, we will perform some filtering by prevalence. If a gene is only expressed in a handful of cells, it is not particularly meaningful as it still brings down the averages for all other cells it is not expressed in. For our data we choose to keep only genes which are expressed in 10 or more cells. By using this filter, genes which have zero counts in all cells will effectively be removed.

```{r}
# Sums all TRUE values and returns TRUE if more than 10 TRUE values per gene
keep_genes <- Matrix::rowSums(nonzero) >= 10

# Only keeping those genes expressed in more than 10 cells
filtered_counts <- counts[keep_genes, ]
```

Finally, take those filtered counts and create a new Seurat object for downstream analysis.

```{r}
# Reassign to filtered Seurat object
filtered_seurat <- CreateSeuratObject(filtered_counts, meta.data = filtered_seurat@meta.data)
```


## Reassessing QC metrics
```{r}
# extracting metadata as a dataframe from the filtered_seurat object
filtered_metadata <- filtered_seurat@meta.data
```


```{r}
# Visualize the number of cells in each sample
  filtered_metadata %>% 
  ggplot() +
  geom_bar(aes(sample, fill = sample)) +
  theme_light() +
  theme(plot.title = element_text(hjust = 0.5, face = "bold")) +
  ggtitle("NCells_filtered")

```
After filtering, the number of ctrl and treated cells is
```{r}
nrow(filtered_metadata %>% filter(sample == "ctrl"))
nrow(filtered_metadata %>% filter(sample == "treat"))
```

```{r}
# visualize the number of transcripts (UMIs) per cell
filtered_metadata %>% 
  ggplot(aes(x = nUMI, color = sample, fill = sample)) +
  geom_density(alpha = 2/10) +
  theme_light() + 
  scale_x_log10() + 
  ylab("cell density") + 
  geom_vline(xintercept = 500) +
  theme(plot.title = element_text(hjust = 0.5, face = "bold")) + 
  ggtitle("transcripts per cell after filtering")
```
```{r}
# visualize number of genes (nGene) per cell with a histogram
filtered_metadata %>% 
  ggplot(aes(nGene, color = sample, fill = sample)) +
  geom_density(alpha = 2/10) + 
  theme_light() + 
  scale_x_log10() + 
  geom_vline(xintercept = 300) + 
  theme(plot.title = element_text(hjust = 0.5, face = "bold")) + 
  ggtitle("genes per cell after filtering")
```
```{r}
# visualize number of genes as a boxplot
filtered_metadata %>% 
  ggplot(aes(x = sample, y = log10(nGene), fill = sample)) + 
  geom_boxplot(alpha = 5/10) + 
  theme_light() + 
  theme(plot.title = element_text(hjust = 0.5, face = "bold")) + 
  ggtitle("genes per cell after filtering")
```
```{r}
# scatter plot of UMIs versus genes detected per cell colored by ratio of mitochondrial genes
filtered_metadata %>% 
  ggplot(aes(x = nUMI, y = nGene, color = mitoRatio)) +
  geom_jitter() +
  scale_color_gradient(low = "gray90", high = "black") +
  stat_smooth(method = lm) +
  scale_x_log10() +
  scale_y_log10() +
  theme_light() + 
  facet_wrap(~sample)
```

```{r}
# visualizing mitochondrial genes per cell 
filtered_metadata %>% 
  ggplot(aes(mitoRatio, color = sample, fill = sample)) + 
  geom_density(alpha = 2/10) + 
  theme_light() + 
  scale_x_log10() + 
  geom_vline(xintercept = 0.2)
  theme(plot.title = element_text(hjust = 0.5, face = "bold")) + 
  ggtitle("mitochondrial gene count ratio per cell after filtering")
```

```{r}
# visualizing genes per UMI to assess complexity
filtered_metadata %>% 
  ggplot(aes(log10GenesPerUMI, color = sample, fill = sample)) + 
  geom_density(alpha = 2/10) + 
  theme_light()+
  scale_x_log10() + 
  geom_vline(xintercept = 0.8)
  theme(plot.title = element_text(hjust = 0.5, face = "bold")) + 
  ggtitle("genes per UMI after filtering")
```

## Assigning cell cycle phases
#### Assigning cell cycle phases and scores
To assign each cell a score based on its expression of G2/M and S phase markers, we can use the Seurat function CellCycleScoring(). This function calculates cell cycle phase scores based on canonical markers that required as input.

```{r}
# load cell cycle data; the RData file was downloaded from the seurat github site; the HBC tutorial github data folder does not have any .rda files!
load("./data/cc.genes.updated.2019.rda")

# scoring cells for cell cycle
filtered_phased_seurat <- CellCycleScoring(filtered_seurat,
                                 s.features = cc.genes.updated.2019$s.genes,
                                 g2m.features = cc.genes.updated.2019$g2m.genes)
```

```{r}
# view cell cycle scores and phases assigned to cells
glimpse(filtered_phased_seurat@meta.data)
```


## Using sctransform

### From the vignette on sctransform from Hafemeister & Satija
Using sctransform in Seurat
Christoph Hafemeister & Rahul Satija

Compiled: 2021-06-14

Source: vignettes/sctransform_vignette.Rmd
Biological heterogeneity in single-cell RNA-seq data is often confounded by technical factors including sequencing depth. The number of molecules detected in each cell can vary significantly between cells, even within the same celltype. Interpretation of scRNA-seq data requires effective pre-processing and normalization to remove this technical variability. In Hafemeister and Satija, 2019 we introduce a modeling framework for the normalization and variance stabilization of molecular count data from scRNA-seq experiment. This procedure omits the need for heuristic steps including pseudocount addition or log-transformation and improves common downstream analytical tasks such as variable gene selection, dimensional reduction, and differential expression.

In this vignette, we demonstrate how using sctransform based normalization enables recovering sharper biological distinction compared to log-normalization.

library(Seurat)
library(ggplot2)
library(sctransform)



### Additional note regarding omitting NormalizeData, ScaleData and FindVariableFeatures functions
Apply sctransform normalization

Note that this single command replaces NormalizeData(), ScaleData(), and FindVariableFeatures().
Transformed data will be available in the SCT assay, which is set as the default after running sctransform

### SCTransform direct

We will apply sctransform directly to the filtered_phased_seurat object without normalizing etc.

```{r}
# split seurat object by condition for sctransform
filtered_phased_seurat <- SplitObject(filtered_phased_seurat,split.by = "sample")
filtered_phased_seurat <- filtered_phased_seurat[c("ctrl", "treat")]
```

SCTransform can generate large objects in terms of memory, hence adjust the limit for allowable object sizes within R 
```{r}
options(future.globals.maxSize = 4000 * 1024^2)
```

Iterating sctransform over each object in filtered_phased_seurat

```{r}
 for (i in seq_along(filtered_phased_seurat)){
  filtered_phased_seurat[[i]] = SCTransform(filtered_phased_seurat[[i]]) # we do not regress out mitochondrial gene ratio and cell cycle scores because these are affected in CLL and are biologically meaningful
}
```

checking assays stored in objects
```{r}
filtered_phased_seurat$ctrl@assays
```
```{r}
filtered_phased_seurat$treat@assays
```



## Splitting up ctrl and treat data into separate objects
```{r}
filtered_phased_ctrl_seurat <- filtered_phased_seurat$ctrl
filtered_phased_treat_seurat <- filtered_phased_seurat$treat
```

## Visualizing the data with PCA
```{r}
# PCA
filtered_phased_ctrl_seurat <- RunPCA(object = filtered_phased_ctrl_seurat)
filtered_phased_treat_seurat <- RunPCA(object = filtered_phased_treat_seurat)

# plot PCA
PCAPlot(object = filtered_phased_ctrl_seurat)
PCAPlot(object = filtered_phased_treat_seurat)
```

### Visualizing with UMAP
```{r}
# run UMAP
filtered_phased_ctrl_seurat <- RunUMAP(filtered_phased_ctrl_seurat,
                             dims = 1:40,
                             reduction = "pca")
filtered_phased_treat_seurat <- RunUMAP(filtered_phased_treat_seurat,
                             dims = 1:40,
                             reduction = "pca")
```

```{r}
# plot UMAP
DimPlot(filtered_phased_ctrl_seurat)
DimPlot(filtered_phased_treat_seurat)
```

## Clustering
### Determining number of PCs to use
```{r}
# elbow / scree plot
ElbowPlot(filtered_phased_ctrl_seurat,
          ndims = 50)
ElbowPlot(filtered_phased_treat_seurat,
          ndims = 50)
```
We will use 50 PCs to determine clusters

### k-nearest neighbors, clusters at various resolutions
```{r}
# K-nearest neighbors graph
filtered_phased_ctrl_seurat <- FindNeighbors(object = filtered_phased_ctrl_seurat, dims = 1:50)

filtered_phased_treat_seurat <- FindNeighbors(object = filtered_phased_treat_seurat, dims = 1:50)

#determine clusters at various resolutions
filtered_phased_ctrl_seurat <- FindClusters(object = filtered_phased_ctrl_seurat, resolution = c(0.4, 0.5,0.6, 0.8, 1.0, 1.4))

filtered_phased_treat_seurat <- FindClusters(object = filtered_phased_treat_seurat, resolution = c(0.4, 0.5,0.6, 0.8, 1.0, 1.4))
```

### looking at various resolutions
```{r}
# filtered_phased_ctrl_seurat@meta.data %>% 
#  View()
# filtered_phased_treat_seurat@meta.data %>% 
# View()
```
shows tables of cells sorted into clusters at various resolutions, not too informative, kept if may be useful, but currently commented out

Choosing resolution of 1.4
```{r}
# assigning identity of clusters
Idents(object = filtered_phased_ctrl_seurat) <- "SCT_snn_res.1.4"
Idents(object = filtered_phased_treat_seurat) <- "SCT_snn_res.1.4"
```

We have calculated UMAP earlier, hence we only visualize the clusters
```{r}
# Plot the UMAP
DimPlot(object = filtered_phased_ctrl_seurat,
        reduction = "umap",
        label = TRUE,
        label.size = 6)

DimPlot(object = filtered_phased_treat_seurat,
        reduction = "umap",
        label = TRUE,
        label.size = 6)
```

Choosing resolution of 0.5
```{r}
# assigning identity of clusters
Idents(object = filtered_phased_ctrl_seurat) <- "SCT_snn_res.0.5"
Idents(object = filtered_phased_treat_seurat) <- "SCT_snn_res.0.5"
```

We have calculated UMAP earlier, hence we only visualize the clusters
```{r}
# Plot the UMAP
DimPlot(object = filtered_phased_ctrl_seurat,
        reduction = "umap",
        label = TRUE,
        label.size = 6)

DimPlot(object = filtered_phased_treat_seurat,
        reduction = "umap",
        label = TRUE,
        label.size = 6)
```

## Exploratory analysis of clusters
### Distribution of cells per cluster in each condition
```{r}
# extracting identity and sample information to determine cells per cluster in each condition
n_cells_ctrl <- FetchData(filtered_phased_ctrl_seurat,
                     vars = c("ident", "orig.ident")) %>% 
  count(ident, orig.ident) %>% 
  spread(ident, n)

n_cells_ctrl

n_cells_treat <- FetchData(filtered_phased_treat_seurat,
                     vars = c("ident", "orig.ident")) %>% 
  count(ident, orig.ident) %>% 
  spread(ident, n)

n_cells_treat
```

```{r}
# visualizing by count and condition
DimPlot(object = integrated_filtered_phased_noMito_seurat,
        split.by = "sample",
        reduction = "umap",
        label = TRUE) +
  NoLegend()
```
### Segregation of clusters by other uninteresting sources of variation
Exploring additional metrics, such as the number of UMIs and genes per cell, S-phase and G2M-phase markers, and mitochondrial gene expression by UMAP. Looking at the individual S and G2M scores can give us additional information to checking the phase as we did previously.
```{r}
# Plot metrics present in seurat_integrated@meta.data
metrics <- c("nUMI", "nGene","S.Score", "G2M.Score", "mitoRatio")

FeaturePlot(filtered_phased_ctrl_seurat,
            reduction = "umap",
            features = metrics,
            pt.size = 0.4,
            order = TRUE,
            min.cutoff = "q10",
            label = TRUE)

FeaturePlot(filtered_phased_treat_seurat,
            reduction = "umap",
            features = metrics,
            pt.size = 0.4,
            order = TRUE,
            min.cutoff = "q10",
            label = TRUE)
```
Some indication that clusters in both conditions may be influenced by these factors?


### Exploring PCs that drive different clusters
The hope is that defined PCs separate the cell types well. To do this, we extract UMAP coordinate information for the cells along with the corresponding scores for each of the first 16 PCs to view by UMAP
```{r}
# defining information of interest in the seurat object
columns <- c(paste0("PC_", 1:16),
             "ident",
             "UMAP_1", "UMAP_2")
# extracting this information from the seurat object
pc_data_ctrl <- FetchData(filtered_phased_ctrl_seurat, 
                     vars = columns)

pc_data_treat <- FetchData(filtered_phased_treat_seurat, 
                     vars = columns)
```

In the UMAP plots below, cells are colored by their score for each of first 16 PCs
```{r}
# adding cluster label to center of cluster on UMAP
umap_label_ctrl <- FetchData(filtered_phased_ctrl_seurat,
                        vars = c("ident", "UMAP_1", "UMAP_2")) %>% 
  group_by(ident) %>% 
  summarize(x = mean(UMAP_1), y = mean(UMAP_2))

umap_label_treat <- FetchData(filtered_phased_treat_seurat,
                        vars = c("ident", "UMAP_1", "UMAP_2")) %>% 
  group_by(ident) %>% 
  summarize(x = mean(UMAP_1), y = mean(UMAP_2))

# plotting a UMAP plot of each of first 16 PCs
map(paste0("PC_", 1:16), function(pc){
  ggplot(pc_data_ctrl,
         aes(UMAP_1, UMAP_2)) +
    geom_point(aes_string(color = pc),
               alpha = 0.7) + 
    scale_color_gradient(guide = "none",
                         low = "grey90",
                         high = "blue") + 
    geom_text(data = umap_label_ctrl,
              aes(label = ident, x, y)) + 
    ggtitle(pc)
}) %>% 
  plot_grid(plotlist = .) 

map(paste0("PC_", 1:16), function(pc){
  ggplot(pc_data_treat,
         aes(UMAP_1, UMAP_2)) +
    geom_point(aes_string(color = pc),
               alpha = 0.7) + 
    scale_color_gradient(guide = "none",
                         low = "grey90",
                         high = "blue") + 
    geom_text(data = umap_label_treat,
              aes(label = ident, x, y)) + 
    ggtitle(pc)
}) %>% 
  plot_grid(plotlist = .) 
```
In the "treat" data, cluster 5 shows very different values on first 3 PCs compared to others

Examine PCA results
```{r}
print(filtered_phased_ctrl_seurat[["pca"]], dims = 1:5, nfeatures = 5)
print(filtered_phased_treat_seurat[["pca"]], dims = 1:5, nfeatures = 5)
```


## Visualizing known cell-type markers

```{r}
# select the RNA counts slot to be default assay
DefaultAssay(filtered_phased_ctrl_seurat) <- "RNA"
DefaultAssay(filtered_phased_treat_seurat) <- "RNA"
# Normalize RNA data for visualization
filtered_phased_ctrl_seurat <- NormalizeData(filtered_phased_ctrl_seurat, verbose = FALSE)

filtered_phased_treat_seurat <- NormalizeData(filtered_phased_treat_seurat, verbose = FALSE)
```

### Plotting known immune and CLL markers
```{r}
# DotPlot using most variable features and other known markers
DotPlot(filtered_phased_ctrl_seurat,features = c("CD79A","MS4A1", "CD19","CCR7"))

DotPlot(filtered_phased_treat_seurat,features = c("CD79A","MS4A1", "CD19","CCR7"))
```
Major T cell (CD3D, CD4, CD8A and IL7R) markers were not found in the data

## Finding markers differentially expressed between B-lymphocyte like cells and other clusters

### DE between cluster5 and 0,1,2 for 'ctrl' data
```{r}
# DE between 0,1,2 and cluster 5
ctrl_cluster5cells <- FindMarkers(filtered_phased_ctrl_seurat,
                             ident.1 = 5,
                             ident.2 = c(0,1,2))
# add gene symbols to table
ctrl_cluster5cells <- ctrl_cluster5cells %>%
  rownames_to_column(var = "gene") %>%
  left_join(y = unique(annotations[, c("gene_name", "description")]),
             by = c("gene" = "gene_name"))
# reorder columns
ctrl_cluster5cells <- ctrl_cluster5cells[ ,c(1, 3:5, 2, 6:7)]

# sort by padj value
ctrl_cluster5cells <-  ctrl_cluster5cells %>% 
  arrange(p_val_adj)

View(ctrl_cluster5cells)

# sort by difference in percent cells in each condition expressing gene
DE_ctrl_cluster5cells <-  ctrl_cluster5cells %>% 
  mutate(pct_clust5_cells_minus_pct_other = pct.2-pct.1) %>% 
  arrange(desc(pct_clust5_cells_minus_pct_other))

View(DE_ctrl_cluster5cells)
```

### DE between cluster4 & cluster5 and 0,1,2 for 'treat' data
```{r}
# DE between 0,1,2 and cluster 4
treatd_cluster4cells <- FindMarkers(filtered_phased_treat_seurat,
                             ident.1 = 4,
                             ident.2 = c(0,1,2))
# add gene symbols to table
treatd_cluster4cells <- treatd_cluster4cells %>%
  rownames_to_column(var = "gene") %>%
  left_join(y = unique(annotations[, c("gene_name", "description")]),
             by = c("gene" = "gene_name"))
# reorder columns
treatd_cluster4cells <- treatd_cluster4cells[ ,c(1, 3:5, 2, 6:7)]

# sort by padj value
treatd_cluster4cells <-  treatd_cluster4cells %>% 
  arrange(p_val_adj)

View(treatd_cluster4cells)

# sort by difference in percent cells in each condition expressing gene
DE_treatd_cluster4cells <-  treatd_cluster4cells %>% 
  mutate(pct_clust4_cells_minus_pct_other = pct.2-pct.1) %>% 
  arrange(desc(pct_clust4_cells_minus_pct_other))

View(DE_treatd_cluster4cells)

```

### DE between cluster 5 and clusters 0,1,2 of 'treat' data
```{r}
# DE between 0,1,2 and cluster 5 of treated cells
treatd_cluster5cells <- FindMarkers(filtered_phased_treat_seurat,
                             ident.1 = 5,
                             ident.2 = c(0,1,2))
# add gene symbols to table
treatd_cluster5cells <- treatd_cluster5cells %>%
  rownames_to_column(var = "gene") %>%
  left_join(y = unique(annotations[, c("gene_name", "description")]),
             by = c("gene" = "gene_name"))
# reorder columns
treatd_cluster5cells <- treatd_cluster5cells[ ,c(1, 3:5, 2, 6:7)]

# sort by padj value
treatd_cluster5cells <-  treatd_cluster5cells %>% 
  arrange(p_val_adj)

View(treatd_cluster5cells)

# sort by difference in percent cells in each condition expressing gene
DE_treatd_cluster5cells <-  treatd_cluster5cells %>% 
  mutate(pct_treat_clust5_cells_minus_pct_other = pct.2-pct.1) %>% 
  arrange(desc(pct_treat_clust5_cells_minus_pct_other))

View(DE_treatd_cluster5cells)
```



