---
title: "Seurat_guided_analysis_tutorial"
author: "ASG"
date: "7/12/2021"
output: github_document
toc: true
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Downloading & reading in the data
Downloaded the peripheral blood mononuclear cell dataset from
https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz
There are 2,700 single cells that were sequenced on the Illumina NextSeq 500.

tar-xvzfed the dataset in the scRNASeq_SR1 folder in a separate folder called pbmc.

The Read10X() function reads in the output of the cellranger pipeline from 10X, returning a unique molecular identified (UMI) count matrix. The values in this matrix represent the number of molecules for each feature (i.e. gene; row) that are detected in each cell (column).

We next use the count matrix to create a Seurat object. The object serves as a container that contains both data (like the count matrix) and analysis (like PCA, or clustering results) for a single-cell dataset. For a technical discussion of the Seurat object structure, check out our GitHub Wiki. For example, the count matrix is stored in pbmc[["RNA"]]@counts.

```{r}
library(dplyr)
library(Seurat)
library(patchwork)
(.packages())
```
```{r}
# load the pbmc data
pbmc.data <- Read10X("./data/pbmc/filtered_gene_bc_matrices/hg19")
# initialize the Seurat object with the raw (non-normalized) data
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, min.features = 200)
pbmc
```

